controllers/crds/cninode_controller.go

```diff
@@ -276,17 +251,60 @@ func (r *CNINodeReconciler) createCNINodeFromObj(ctx context.Context, newCNINode
 		})
 }
 
-func (r *CNINodeReconciler) GetNodeID(ctx context.Context, nodeName string) (string, error) {
-	node := &v1.Node{}
-	err := r.Client.Get(ctx, types.NamespacedName{Name: nodeName}, node)
-	if err != nil {
-		return "", fmt.Errorf("failed to get the node object %s node, err: %s", nodeName, err)
+func (r *CNINodeReconciler) ensureTagsAndLabels(cniNode *v1alpha1.CNINode, node *v1.Node, nodeFound bool) (bool, error) {
+	shouldPatch := false
+	var err error
+	if cniNode.Spec.Tags == nil {
+		cniNode.Spec.Tags = make(map[string]string)
 	}
-	if node.Spec.ProviderID == "" {
-		return "", fmt.Errorf("provider ID is not set for node %s", nodeName)
+	// add cluster name tag if it does not exist
+	if cniNode.Spec.Tags[config.VPCCNIClusterNameKey] != r.clusterName {
+		cniNode.Spec.Tags[config.VPCCNIClusterNameKey] = r.clusterName
+		shouldPatch = true
 	}
-	if idx := strings.LastIndex(node.Spec.ProviderID, "/"); idx != -1 && idx < len(node.Spec.ProviderID)-1 {
-		return node.Spec.ProviderID[idx+1:], nil
+	if nodeFound {
+		var nodeID string
+		nodeID, err = utils.GetNodeID(node)
+
+		if cniNode.Spec.Tags[config.NetworkInterfaceNodeIDKey] != nodeID {
+			cniNode.Spec.Tags[config.NetworkInterfaceNodeIDKey] = nodeID
+			shouldPatch = true
+		}
+
+		// add node label if it does not exist
+		if cniNode.ObjectMeta.Labels == nil {
+			cniNode.ObjectMeta.Labels = make(map[string]string)
+		}
+		if cniNode.ObjectMeta.Labels[config.NodeLabelOS] != node.ObjectMeta.Labels[config.NodeLabelOS] {
+			cniNode.ObjectMeta.Labels[config.NodeLabelOS] = node.ObjectMeta.Labels[config.NodeLabelOS]
+			shouldPatch = true
+		}
+	}
+	return shouldPatch, err
+}
+
+func (r *CNINodeReconciler) ensureFinalizer(cniNode *v1alpha1.CNINode) bool {
+	shouldPatch := false
```

Comment:
We don't need this variable, do we? We just return `controllerutil.AddFinalizer()` which returns a boolean.